lib/ClangImporter/ImportName.cpp

```diff
@@ -576,12 +574,96 @@ determineCtorInitializerKind(const clang::ObjCMethodDecl *method) {
   return None;
 }
 
+template <typename A>
+static bool matchesVersion(A *versionedAttr, ImportNameVersion version) {
+  clang::VersionTuple attrVersion = versionedAttr->getVersion();
+  if (attrVersion.empty()) {
+    // FIXME: This is brittle. The empty version represents "newest", as in
+    // "newer than the current version", but ImportNameVersion::Swift4 is
+    // a specific new version.
+    return version == ImportNameVersion::Swift4;
+  } else {
+    return attrVersion.getMajor() == majorVersionNumberForNameVersion(version);
+  }
+}
+
+const clang::SwiftNameAttr *
+importer::findSwiftNameAttr(const clang::Decl *decl,
+                            ImportNameVersion version) {
+  if (version == ImportNameVersion::Raw)
+    return nullptr;
+
+  // Handle versioned API notes for Swift 3 and later.
+  if (version != ImportNameVersion::Swift2) {
+    for (auto *attr : decl->attrs()) {
+      if (auto *versionedAttr = dyn_cast<clang::SwiftVersionedAttr>(attr)) {
+        if (!matchesVersion(versionedAttr, version))
+          continue;
+        if (auto *added =
+              dyn_cast<clang::SwiftNameAttr>(versionedAttr->getAttrToAdd())) {
+          return added;
+        }
+      }
+
+      if (auto *removeAttr = dyn_cast<clang::SwiftVersionedRemovalAttr>(attr)) {
+        if (!matchesVersion(removeAttr, version))
+          continue;
+        if (removeAttr->getAttrKindToRemove() == clang::attr::SwiftName)
+          return nullptr;
+      }
+    }
+  }
```

Comment:
Would it simplify logic to put in a `return decl->getAttr<clang::SwiftNameAttr>()'` before this last brace, then you can drop the check later on?